src/Generator/Generators/CSharp/CSharpMarshal.cs

```diff
@@ -382,16 +385,26 @@ public override bool VisitArrayType(ArrayType array, TypeQualifiers quals)
             if (!VisitType(array, quals))
                 return false;
 
+            Class @class;
             switch (array.SizeType)
             {
                 case ArrayType.ArraySize.Constant:
                     var supportBefore = Context.SupportBefore;
                     supportBefore.WriteLine("if ({0} != null)", Context.ArgName);
                     supportBefore.WriteStartBraceIndent();
+                    if (array.Type.Desugar().TryGetClass(out @class) && @class.IsRefType)
+                    {
+                        supportBefore.WriteLine("if (value.Length != {0})", array.Size);
+                        supportBefore.WriteLineIndent("throw new ArrayTypeMismatchException(\"The provided array's dimensions doesn't match the required size.\");");
```

Comment:
Out of range does not imply which side of the range the value is out of. For example, if your List<int> has a size of 5 and you request the value at index -1, you'd still get an IndexOutOfRangeException even though -1 < 5.